Prow pipeline controller fixes and improvements

cjwagner opened the pull request:
This PR makes some fixes and improvements to Prow's pipeline controller:
- Fix pipeline git resource usage pattern.
  - Previously the git resource got the ProwJob name which makes it impossible to reference from `pipeline_run_spec` in config since the ProwJob name is random and different for every run.
  - Now 'magic' reference values in `pipeline_run_spec` git resource references like `PROW_IMPLICIT_GIT_REF` will be substituted for the actual resource names by the controller before the pipelinerun is created.
  - Also adds similar support for `extra_refs` substitution and pipelineresource creation.
  - Avoids creating or referencing pipeline git resources if they are not referenced in the `pipeline_run_spec` which resulted in invalid pipelineruns.
- Provide all standard downwardapi env vars as pipeline parameters.
- Ignore ProwJobs for different agents.
- Fixed state transitioning so ProwJobs are considered pending if a pipelinerun exists.
  - ProwJobs should be considered pending as soon as the execution unit (pod, build, pipelinerun) is created for consistency with the other controllers. If this change is questionable we should consider updating the other controllers to match the pipeline controller's behavior.
  - I noticed this because I was seeing the pipeline controller reliably log transitions from `triggered` -> `successful` without seeing `pending`.
- Deleted unused function.
- Fixed some ineffective test cases.

Alternative implementation of https://github.com/kubernetes/test-infra/pull/13083

We'll still need to implement proper garbage collection of the git pipelineresources (probably using owner references) in a follow up PR.

/assign @ccojocar @vdemeester @wbrefvem @stevekuznetsov 
/cc @bobcatfish

vdemeester commented on prow/cmd/pipeline/controller.go:
```diff
@@ -586,31 +591,72 @@ func makePipelineGitResource(pj prowjobv1.ProwJob) *pipelinev1alpha1.PipelineRes
+	}
+	for _, key := range sets.StringKeySet(env).List() {
+		val := env[key]
+		// TODO: make this handle existing values/substitutions.
```
@cjwagner what do you mean by that ?

cjwagner replied:
As is this just appends the Prow parameters to the list without regard for existing parameters. If a Prow parameter is explicitly specified by the user we probably will want to error (at config validation time as well) or replace the value. I'm not sure which option is more appropriate. WDYT?

anarcher commented:
IMHO, It should be optional that Prow pipeline controller deletes PipelineRun. Because that In some cases It would like to keep tekton PipelineRuns And prow pipeline controller retries  (new)PipelineRun is optional too. :->

anarcher commented:
Sometimes, I would like that  PipelineRun name  is from `{Pipeline.Name}-{ProwJob.Status.BuildID}`.  IMHO, This name is more readable from tekton's view. :-)

cjwagner commented:
> IMHO, It should be optional that Prow pipeline controller deletes PipelineRun. Because that In some cases It would like to keep tekton PipelineRuns. And prow pipeline controller retries (new)PipelineRun can be optional too. :->

They are currently kept around for the lifetime of the ProwJob (1 week by default). We need to delete completed pipelineruns so that their associated resources (namely the containers) can be deleted. We do the same thing for pods and knative builds. If anything we will actually want to delete the completed pipelineruns more aggressively (like we do with pods) and just keep the ProwJob resources around. 
It sounds like you may want to increase the `max_prowjob_age` for your Prow instance if you want pipelineruns to persist for more than a week:
https://github.com/kubernetes/test-infra/blob/f3f5f552f40d5d614f4c2d55d7337db9bf3e4ec3/prow/config.yaml#L23

> Sometimes, I would like that PipelineRun name is from `{Pipeline.Name}-{ProwJob.Status.BuildID}`. IMHO, This name is more readable from tekton's dashboard. :-)

The pipelinerun matches the ProwJob name because there is a 1-1 relationship. We do the same thing for pods and knative builds. It makes debugging easier and is also widely assumed. 
We could generate ProwJob names based on a scheme like this to provide human readable names for both, but we can't use the build ID because that hasn't been determined when the job is first triggered (although we could possibly change that too).

While these are good things to consider, my PR doesn't change either of these things so they shouldn't block this.
PTAL

anarcher commented:
I agree that @cjwagner 's comments 👍 

> We need to delete completed pipelineruns so that their associated resources (namely the containers) can be deleted.

When the controller does that, `pipeline_run_spec` git resource references (`PROW_IMPLICIT_GIT_REF`)  are deleted?

anasinnyk commented:
Hi @cjwagner , 
In new tekton 0.5.0 we have a PullRequest resource, maybe it's a better solution than git resource because Prow working with concrete pull request and not with commit only.

For example, we have a simple linter, but we want to check all our project, we want to check the only diff. If we check each commit the last commit can pass it, but whole pull request can contain incorrect linting. And right now we set PR status after last job running.

What do you think?

stevekuznetsov commented:
Does it matter if some transient state in the middle of a chain of commits doesn't match? I would not like to use "pull request" if it means we lose the ability to specifically tie a job to a commit in a PR and a commit in the branch that is being targeted.

stevekuznetsov commented:
Does it matter if some transient state in the middle of a chain of commits doesn't match? I would not like to use "pull request" if it means we lose the ability to specifically tie a job to a commit in a PR and a commit in the branch that is being targeted.

cjwagner commented:
@anarcher 
> When the controller does that, `pipeline_run_spec` git resource references (`PROW_IMPLICIT_GIT_REF`) are deleted?

Not yet, but this is mentioned in the OP: "We'll still need to implement proper garbage collection of the git pipelineresources (probably using owner references) in a follow up PR."

@anasinnyk 
PullRequest resources wouldn't make sense for postsubmit refs or extra refs so we'd still need normal git PipelineResources either way. We can already determine which files were changed by a PR using git or GitHub so we can already "check only the diff" of the PR without needing to using a PullRequest resource. If there are more benefits to using a PullRequest resource and none of the downsides Steve is concerned about it could make sense, but as of now it seems like extra complexity without benefit.

vdemeester commented:
Agreeing with @cjwagner on the PullRequest resource. The main benefit for the PullRequest resource is to be able interact with the API (GitHub right now), to add comment or custom checks. And I really think this is too early to integrate with it in Prow (nor it's this PR role) :angel:

anasinnyk commented:
@cjwagner @vdemeester thanks for explenation

cjwagner commented:
:man_facepalming: I added these commands in an edit instead of a new message which I know doesn't work.
/assign @ccojocar @wbrefvem @stevekuznetsov
/cc @bobcatfish

Rebased, please review.

JoelSpeed commented on prow/config/config.go:
```diff
@@ -1295,6 +1302,52 @@ func resolvePresets(name string, labels map[string]string, spec *v1.PodSpec, bui
+		if len(match) != 2 {
+			continue
+		}
+		if len(match[1]) > 1 && match[1] == "0" {
```
I'm not sure this is correct, we check `match[1]` is greater in length than 1 character and then check if `match[1] == "0"`, a single `0`?

I had a play in the playground and can pass `PROW_EXTRA_GIT_REF_001` through this code without being invalidated, I think we are trying to catch that?
https://play.golang.org/p/U3Tm2v6dGkc

cjwagner replied:
Good catch! I meant for this to check the first character, not the entire string.
I fixed this and added tests for this function.
